pyanaconda/modules/payloads/payload/rpm_ostree/installation.py

```diff
@@ -0,0 +1,342 @@
+#
+# Copyright (C) 2020 Red Hat, Inc.
+#
+# This copyrighted material is made available to anyone wishing to use,
+# modify, copy, or redistribute it subject to the terms and conditions of
+# the GNU General Public License v.2, or (at your option) any later version.
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY expressed or implied, including the implied warranties of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
+# Public License for more details.  You should have received a copy of the
+# GNU General Public License along with this program; if not, write to the
+# Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+# 02110-1301, USA.  Any Red Hat trademarks that are incorporated in the
+# source code or documentation are not subject to the GNU General Public
+# License and may only be used or replicated with the express permission of
+# Red Hat, Inc.
+#
+import os
+
+from pyanaconda.core.configuration.anaconda import conf
+from pyanaconda.core.glib import Variant
+from pyanaconda.core.util import mkdirChain
+from pyanaconda.modules.common.task import Task
+from pyanaconda.modules.common.constants.objects import DEVICE_TREE, BOOTLOADER
+from pyanaconda.modules.common.constants.services import STORAGE
+from pyanaconda.modules.payloads.payload.rpm_ostree.utils import safe_exec_with_redirect
+from pyanaconda.modules.common.structures.storage import DeviceData
+
+from pyanaconda.anaconda_loggers import get_module_logger
+log = get_module_logger(__name__)
+
+
+class PrepareOSTreeMountTargetsTask(Task):
+    """Task to prepare OSTree mount targets."""
+
+    def __init__(self, source_config):
+        super().__init__()
+        self._source_config = source_config
+        self._internal_mounts = []
+
+    @property
+    def name(self):
+        return "Prepare OSTree mount targets"
+
+    def _setup_internal_bindmount(self, src, dest=None,
+                                  src_physical=True,
+                                  bind_ro=False,
+                                  recurse=True):
+        """Internal API for setting up bind mounts between the physical root and
+           sysroot, also ensures we track them in self._internal_mounts so we can
+           cleanly unmount them.
+
+           :param src: Source path, will be prefixed with physical or sysroot
+           :param dest: Destination, will be prefixed with sysroot (defaults to same as src)
+           :param src_physical: Prefix src with physical root
+           :param bind_ro: Make mount read-only
+           :param recurse: Use --rbind to recurse, otherwise plain --bind
+        """
+        # Default to the same basename
+        if dest is None:
+            dest = src
+        # Almost all of our mounts go from physical to sysroot
+        if src_physical:
+            src = conf.target.physical_root + src
+        else:
+            src = conf.target.system_root + src
```

Comment:
We usually specify the sysroot via a parameter of the task. Sometimes it simplifies testing, because you can set the sysroot to a temporary directory. You might want to do the same for the physical root in that case.